### ISSUE TITLE
Clearer Error message when Module's __call__ is not implemented

### ISSUE THREAD

**Issue by MiladInk** [google/flax#975]
Hi All!

It has been just a few days that I have been using Flax. Once, I forgot to write the `__call__` of a function of my neural net class for a very funny reason. 

This is what I wrote:

```
class NN(nn.Module):
  features: Sequence[int]
  
  def setup(self):
    self.layers = [nn.Dense(feat) for feat in self.features]
```

Flax raises a warning in this settings that I did not find informational:

`ValueError: parent must be None, Module or Scope`

It does not point out that I should just write the `__call__` function.

 I am wondering if the message could be clearer as I think it may help the people who just started using Jax.

Thank you all for the great code!

**Comment by marcvanzee** [comment]
Hi @MiladInk, thanks a lot, this input is very valuable to us! I'm working on an effort to improve error messages in Flax this quarter, so I'll include this issue as well. As a sneak peak: we basically want to make sure that all error thrown are error classes that can be found here: https://flax.readthedocs.io/en/latest/flax.errors.html

**Comment by marcvanzee** [comment]
Hi @MiladInk, I resolved your issue in #1084. I also added a test there showing the desired behavior.
